File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/inheritance/class_delegation/class_delegation.kt
// KIND: STANDALONE
// FREE_COMPILER_ARGS: -opt-in=kotlin.native.internal.InternalForKotlinNative
// MODULE: Inheritance
// FILE: class_delegation.kt

interface DelegatedContract {
    fun delegatedValue(): String
}

private class DelegatedImplementation : DelegatedContract {
    override fun delegatedValue(): String = "delegated-implementation"
}

open class DelegatingBase : DelegatedContract by DelegatedImplementation() {
    open fun localValue(): String = "delegating-base"
}

class DelegatingStorage {
    private var stored: DelegatingBase? = null

    fun store(value: DelegatingBase) {
        stored = value
    }

    fun retrieve(): DelegatingBase? = stored
}

fun callDelegatedValue(value: DelegatedContract): String = value.delegatedValue()
fun callDelegatingLocal(value: DelegatingBase): String = value.localValue()

File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/inheritance/class_delegation/class_delegation.swift
import Inheritance
import Testing

@Test
func kotlinClassDelegationSurvivesSwiftSubclassing() throws {
    class SwiftDelegatingLeaf: DelegatingBase {
        override func localValue() -> String { "swift-local" }
    }

    let value = SwiftDelegatingLeaf()
    let storage = DelegatingStorage()
    storage.store(value: value)
    let retrieved = try #require(storage.retrieve())
    #expect(retrieved === value)
    #expect(callDelegatedValue(value: value) == "delegated-implementation")
    #expect(callDelegatingLocal(value: value) == "swift-local")
}
